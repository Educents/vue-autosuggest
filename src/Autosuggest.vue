<template>
    <div id="autosuggest">
        <input class="form-control"
               name="q"
               type="text"
               maxlength="256"
               autocomplete="off"
               v-model="searchInput"
               :class="[isOpen ? 'autosuggest__input-open' : '', inputProps['class']]"
               @keydown="handleKeyStroke"
               @click="onClick"
               v-bind="inputProps"
               aria-autosuggest="list"
               aria-owns="autosuggest__results"
               :aria-activedescendant="isOpen ? `autosuggest__results--item-${currentIndex}` : ''"
               :aria-haspopup="isOpen"
        />
        <div class="autosuggest__results-container">
                <div
                    role="listbox" 
                    class="autosuggest__results" 
                    aria-labelledby="autosuggest"
                    v-if="getSize() > 0 && !loading"
                    >
                    <component v-for="(cs, key) in this.computedSections" 
                        :is="cs.name" :section="cs" :ref="getSectionRef(key)" :key="getSectionRef(key)" :updateCurrentIndex="updateCurrentIndex" :searchInput="searchInput"></component>
                </div>
        </div>
    </div>
</template>

<script>

import DefaultSection from './parts/DefaultSection.vue';
import UrlSection from './parts/UrlSection.vue';

export default {
        name: 'autosuggest',
        components: {
            DefaultSection,
            UrlSection
        },
        props: {
            inputProps: {
                id: {
                    type: String,
                    default: 'autosuggest__input'
                },
                onInputChange: {
                    type: Function,
                    required: true
                },
                initialValue: {
                    type: String,
                    default: ""
                },
                placeholder: {
                    type: String,
                    required: true
                },
                onClick: {
                    type: Function,
                    required: false
                }
            },
            limit: {
                type: Number,
                required: false,
                default: Infinity
            },
            suggestions: {
                type: Array,
                required: true,
                default: []
            },
            shouldRenderSuggestions: {
                type: Function,
                required: false,
                default: () => {
                    return true;
                }
            },
            resultItemKey: {
                type: String,
                required: false
            },
            sectionConfigs: {
                type: Object,
                required: false
            }
        },
        data: () => ({
            searchInput: '',
            searchInputOriginal: '',
            currentIndex: null,
            currentItem: null,
            loading: false, /** Helps with making sure the dropdown doesn't stay open after certain actions */
            didSelectFromOptions: false,
            computedSections: [],
            computedSize: 0,
            onSelected: function(){
                if (this.currentItem && this.sectionConfigs[this.currentItem.type]) {
                    this.sectionConfigs[this.currentItem.type].onSelected(this.currentItem);
                }
            }
        }),
        computed: {
            isOpen() {
                return this.shouldRenderSuggestions() && !this.loading;
            },
            suggestionsFiltered(){
                return this.suggestions.slice(0, this.limit);
            }
        },
        created() {
        },
        methods: {
            getSectionRef(i) {
                return 'computed_section_' + i;
            },
            getSize() {
                return this.computedSize;
            },
            getItemByIndex(index) {
                let obj = false;
                if (!index) return obj;
                for (var i = 0; i < this.computedSections.length; i++) {
                    if (index >= this.computedSections[i].start_index && index <= this.computedSections[i].end_index) {
                        let trueIndex = index - this.computedSections[i].start_index;
                        let childSection = this.$refs['computed_section_' + i][0];
                        if (childSection) {
                            obj = {
                                "type": this.computedSections[i].type,
                                "label": childSection.getLabelByIndex(trueIndex),
                                "item": childSection.getItemByIndex(trueIndex)
                            };
                            break;
                        }
                    }
                }

                return obj;
            },
            handleKeyStroke(e) {
                const {keyCode} = e;
                this.loading = false;
                this.didSelectFromOptions = false;
                switch (keyCode) {
                    case 40: // ArrowDown
                    case 38: // ArrowUp
                        e.preventDefault();
                        if (this.isOpen) {
                            // Determine direction of arrow up/down and determine new currentIndex
                            const direction = keyCode === 40 ? 1 : -1;
                            const newIndex = this.currentIndex + direction;
                            this.setCurrentIndex(newIndex, this.getSize(), direction);
                            this.didSelectFromOptions = true;
                            if (this.getSize() > 0 && this.currentIndex >= 0) {
                                this.setChangeItem(this.getItemByIndex(this.currentIndex));
                                this.didSelectFromOptions = true;
                            }
                        }
                        break;
                    case 13:  // Enter
                        e.preventDefault();
                        if (keyCode === 229) { // https://github.com/moroshko/react-autosuggest/pull/388
                            break;
                        }
                        this.$nextTick(() => {
                            if (this.getSize() > 0 && this.currentIndex >= 0) {
                                this.setChangeItem(this.getItemByIndex(this.currentIndex));
                                this.didSelectFromOptions = true;
                            }
                            this.loading = true;
                            this.$nextTick(() => {
                                this.onSelected(this.didSelectFromOptions);
                            });
                        });
                        break;
                    case 27: // Escape
                        if (this.isOpen) {
                            /* For 'search' input type, make sure the browser doesn't clear the input when Escape is pressed. */
                            this.loading = true;
                            this.currentIndex = null;
                            this.searchInput = this.searchInputOriginal;
                            e.preventDefault();
                        }
                        break;
                }
            },
            setChangeItem(item) {
                if (item) {
                    this.searchInput = item.label;
                    this.currentItem = item;
                }
            },
            updateCurrentIndex(index) {
                this.currentIndex = index;
            },
            onDocumentMouseUp() {
                /** Clicks outside of dropdown to exit */
                if (this.currentIndex === null) {
                    this.loading = this.shouldRenderSuggestions();
                    return;
                }

                /** Selects an item in the dropdown */
                this.loading = true;
                
                this.setChangeItem(this.getItemByIndex(this.currentIndex));
                this.$nextTick(() => {
                    this.onSelected(true);
                })
                
            },
            setCurrentIndex(newIndex, limit = -1, direction) {
                let adjustedValue = newIndex;

                // if we hit the lower limit then stop iterating the index
                if (this.currentIndex === null) {
                    adjustedValue = 0;
                }

                if (this.currentIndex < 0 && direction === 1) {
                    adjustedValue = 0;
                }
                // if we hit the upper limit then just stop iterating the index
                if (newIndex >= limit) {
                    adjustedValue = 0;
                }
                this.currentIndex = adjustedValue;
                
                const element = document.getElementById(`autosuggest__results_item-${this.currentIndex}`);
                const hoverClass = 'autosuggest__results_item-highlighted';
                if(document.querySelector(`.${hoverClass}`)){
                    this.removeClass(document.querySelector(`.${hoverClass}`),hoverClass);
                }
                if(element){
                    this.addClass(element, hoverClass);
                }
            },
            onClick() {
                this.loading = false;
                this.inputProps.onClick();
            },
            
            /** DOM Utilities */
            hasClass(el, className) {
                if (el.classList)
                    return el.classList.contains(className)
                else
                    return !!el.className.match(new RegExp('(\\s|^)' + className + '(\\s|$)'))
            },
            addClass(el, className) {
                if (el.classList)
                    el.classList.add(className)
                else if (!hasClass(el, className)) el.className += " " + className
            },
            removeClass(el, className) {
                if (el.classList){
                    el.classList.remove(className)
                }
                else if (hasClass(el, className)) {
                    var reg = new RegExp('(\\s|^)' + className + '(\\s|$)')
                    el.className = el.className.replace(reg, ' ')
                }
            }
        },
        mounted() {
            document.addEventListener('mouseup', this.onDocumentMouseUp);
            const input = document.getElementById(this.inputProps.id);
            if(input){
                input.value = this.inputProps.initialValue; // set default query, e.g. loaded server side.
            }
        },
        watch: {
            searchInput(newValue, oldValue) {
                this.value = newValue;
                if(!this.didSelectFromOptions){
                    this.searchInputOriginal = this.value;
                    this.currentIndex = null;
                    this.inputProps.onInputChange(newValue);
                }
            },
            suggestions() {
                this.computedSections = [];
                this.computedSize = 0;
                var generateName = function(name) {
                    name = name.toString();
                    return name.charAt(0).toUpperCase() + name.slice(1) + 'Section';
                }

                var getType = function(type) {
                    if (!type) {
                        type = 'default';
                    }

                    return type;
                }

                this.suggestions.forEach(function(section) {
                    var n = generateName(getType(section.type));
                    var lim = this.$options.components[n].getLimit();
                    lim = (section.data.length < lim) ? section.data.length : lim;
                    var obj = {
                        limit: lim,
                        name: n,
                        data: section.data,
                        label: section.label,
                        start_index: this.computedSize,
                        end_index: this.computedSize + lim - 1,
                        type: getType(section.type)
                    };
                    this.computedSections.push(obj);
                    this.computedSize += lim;
                }, this);
            }
        }
    }
</script>