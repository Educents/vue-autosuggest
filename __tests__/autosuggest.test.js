import { mount, shallow } from "vue-test-utils";
import { createRenderer } from "vue-server-renderer";

import Autosuggest from "../src/Autosuggest.vue";

Element.prototype.scrollTo = () => {}  // https://github.com/vuejs/vue-test-utils/issues/319

// Helper to call function x number of times
const times = x => f => {
  if (x > 0) {
    f();
    times(x - 1)(f);
  }
};

describe("Autosuggest", () => {
  const id = `autosuggest__input`;
  const filteredOptions = [
    {
      data: [
        "clifford kits",
        "friendly chemistry",
        "phonics",
        "life of fred",
        "life of fred math",
        "magic school bus",
        "math mammoth light blue",
        "handwriting",
        "math",
        "minecraft",
        "free worksheets",
        "4th grade",
        "snap circuits",
        "bath toys",
        "channies",
        "fred",
        "lego",
        "math life of fred",
        "multiplication",
        "thinking tree"
      ]
    }
  ];

  const defaultProps = {
    suggestions: filteredOptions,
    inputProps: {
      id,
      initialValue: "",
      onClick: () => {},
      onInputChange: () => {},
      placeholder: "Type 'G'"
    },
    sectionConfigs: {
      default: {
        limit: 5,
        onSelected: () => {}
      }
    }
  };


  it("can mount", () => {
    const props = Object.assign({}, defaultProps);
    props.inputProps = Object.assign({}, defaultProps.inputProps);
    props.suggestions = [filteredOptions[0]];

    const wrapper = shallow(Autosuggest, {
      propsData: props
    });

    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err);
      expect(str).toMatchSnapshot();
    });
  });

  
  it("can be readonly", () => {
    const wrapper = mount(Autosuggest, {
      propsData: {
        ...defaultProps,
        readonly: true,
      }
    });

    const input = wrapper.find("input");
    expect(input.attributes().readonly).toBeTruthy();
    
  })

  it("can be readonly", () => {
    const wrapper = mount(Autosuggest, {
      propsData: {
        ...defaultProps,
        readonly: false,
      }
    });

    const input = wrapper.find("input");
    expect(input.attributes().readonly).toBeFalsy();
    
  })

  it("can render suggestions", () => {
    const wrapper = mount(Autosuggest, {
      propsData: defaultProps
    });

    const input = wrapper.find("input");
    expect(input.attributes("id", defaultProps.inputProps.id)).toBeTruthy();

    input.trigger("click");
    wrapper.setData({ searchInput: "G" });
    input.trigger("keydown.down");

    expect(wrapper.findAll(`ul li`).length).toBeLessThanOrEqual(
      defaultProps.sectionConfigs.default.limit
    );

    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) {
        return false;
      }
      expect(str).toMatchSnapshot();
    });
  });

  it("can use escape key to exit", async () => {
    const wrapper = mount(Autosuggest, {
      propsData: defaultProps
    });

    const input = wrapper.find("input");
    input.trigger("click");
    wrapper.setData({ searchInput: "G" });
    input.trigger("keydown.down");

    expect(wrapper.findAll(`ul li`).length).toBeLessThanOrEqual(
      defaultProps.sectionConfigs.default.limit
    );

    input.trigger("keydown.esc");

    expect(wrapper.findAll(`ul li`).length).toEqual(0);

    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) {
        return false;
      }
      expect(str).toMatchSnapshot();
    });
  });

  it("can select from suggestions using keystroke", async () => {
    const wrapper = mount(Autosuggest, {
      propsData: defaultProps,
      attachToDocument: true
    });

    const input = wrapper.find("input");
    input.trigger("click");
    wrapper.setData({ searchInput: "G" });

    times(5)(() => {
      input.trigger("keydown.down");
    });

    times(5)(() => {
      input.trigger("keydown.up");
    });

    input.trigger("keydown.enter");

    await wrapper.vm.$nextTick(() => {});

    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) {
        return false;
      }
      expect(str).toMatchSnapshot();
    });
  });

  it("can click outside document to trigger close", async () => {
    const props = Object.assign({}, defaultProps);

    const wrapper = mount(Autosuggest, {
      propsData: props,
      attachToDocument: true
    });

    wrapper.setData({ searchInput: "G" });

    const input = wrapper.find("input");

    input.trigger("click");
    wrapper.setData({ searchInput: "G" });
    window.document.dispatchEvent(new Event("mouseup"));

    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) {
        return false;
      }
      expect(str).toMatchSnapshot();
    });
  });

  it("can display section header", async () => {
    const props = Object.assign({}, defaultProps);
    props.sectionConfigs = {
      default: {
        label: "Suggestions",
        limit: 5,
        onSelected: () => {}
      }
    };
    const wrapper = mount(Autosuggest, {
      propsData: props,
      attachToDocument: true
    });

    wrapper.setData({ searchInput: "G" });

    const input = wrapper.find("input");

    input.trigger("click");
    wrapper.setData({ searchInput: "G" });
    expect(wrapper.find("ul li:nth-child(1)").element.innerHTML).toBe(
      props.sectionConfigs.default.label
    );
    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) {
        return false;
      }
      expect(str).toMatchSnapshot();
    });
  });

  it("is aria complete", async () => {
    const wrapper = mount(Autosuggest, {
      propsData: defaultProps
    });

    const input = wrapper.find("input");
    expect(input.attributes('aria-autosuggest','list')).toBeTruthy();
    expect(input.attributes('aria-activedescendant', '')).toBeTruthy();
    expect(input.attributes('aria-owns', 'autosuggest__results')).toBeTruthy();   
    
    // TODO: Make sure aria-completeness is actually 2legit2quit.
    
    input.trigger("click");
    wrapper.setData({ searchInput: "G" });

    expect(input.attributes('aria-haspopup', 'true')).toBeTruthy();
    
    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) {
        return false;
      }
      expect(str).toMatchSnapshot();
    });
  });

  it("can render simplest component with single onSelected", async () => {
    const props = Object.assign({}, defaultProps);
    props.inputProps = Object.assign({}, defaultProps.inputProps);
    props.inputProps.class = "cool-class";
    props.suggestions = filteredOptions;

    delete props.suggestions[0].name; // ensure empty component name is OK
    delete props.sectionConfigs; // ensure empty sectionConfigs is OK
    delete props.inputProps.onClick; // ensure empty onClick is OK

    props.onSelected = () => {};

    const wrapper = mount(Autosuggest, {
      propsData: props,
      attachToDocument: true
    });

    const input = wrapper.find("input");
    input.trigger("click");
    wrapper.setData({ searchInput: "G" });
    
    times(3)(() => {
        input.trigger("keydown.down");
    });
  
    input.trigger("keydown.enter");
    wrapper.find('li').trigger("mouseover");
    wrapper.find('li').trigger("mouseenter");
    wrapper.find('li').trigger("mouseleave");

    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) {
        return false;
      }
      expect(str).toMatchSnapshot();
    });
  });
});

/**
 * **Force** update until vue-test-utils is out of beta
 * @param {*} wrapper mounted Vue component
 */
function runWatchers(wrapper) {
  wrapper.vm._watchers.forEach(watcher => {
    watcher.run();
  });
}
