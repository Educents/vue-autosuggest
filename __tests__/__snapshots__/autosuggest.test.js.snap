// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autosuggest can click outside document to trigger close 1`] = `

<div id="autosuggest-9"
     data-server-rendered="true"
>
  <input type="text"
         autocomplete="off"
         role="combobox"
         aria-autocomplete="list"
         aria-owns="autosuggest__results"
         aria-activedescendant
         aria-haspopup="false"
         aria-expanded="false"
         id="autosuggest__input"
         initialvalue
         oninputchange="function onInputChange() {}"
         placeholder="Type 'G'"
         name="q"
         value="G"
         class="form-control"
  >
  <div class="autosuggest__results-container">
  </div>
</div>

`;

exports[`Autosuggest can display section header 1`] = `

<div id="autosuggest-12"
     data-server-rendered="true"
>
  <input type="text"
         autocomplete="off"
         role="combobox"
         aria-autocomplete="list"
         aria-owns="autosuggest__results"
         aria-activedescendant
         aria-haspopup="true"
         aria-expanded="true"
         id="autosuggest__input"
         initialvalue
         oninputchange="function onInputChange() {}"
         placeholder="Type 'G'"
         name="q"
         value="G"
         class="form-control autosuggest__input-open"
  >
  <div class="autosuggest__results-container">
    <div aria-labelledby="autosuggest-12"
         class="autosuggest__results"
    >
      <ul role="listbox"
          aria-labelledby="autosuggest"
      >
        <li class="autosuggest__results_title autosuggest__results_title_default">
          Suggestions
        </li>
        <li role="option"
            data-suggestion-index="0"
            data-section-name="default"
            id="autosuggest__results_item-0"
            class="autosuggest__results_item"
        >
          clifford kits
        </li>
        <li role="option"
            data-suggestion-index="1"
            data-section-name="default"
            id="autosuggest__results_item-1"
            class="autosuggest__results_item"
        >
          friendly chemistry
        </li>
        <li role="option"
            data-suggestion-index="2"
            data-section-name="default"
            id="autosuggest__results_item-2"
            class="autosuggest__results_item"
        >
          phonics
        </li>
        <li role="option"
            data-suggestion-index="3"
            data-section-name="default"
            id="autosuggest__results_item-3"
            class="autosuggest__results_item"
        >
          life of fred
        </li>
        <li role="option"
            data-suggestion-index="4"
            data-section-name="default"
            id="autosuggest__results_item-4"
            class="autosuggest__results_item"
        >
          life of fred math
        </li>
      </ul>
    </div>
  </div>
</div>

`;

exports[`Autosuggest can mount 1`] = `

<div id="autosuggest-0"
     data-server-rendered="true"
>
  <input type="text"
         autocomplete="off"
         role="combobox"
         aria-autocomplete="list"
         aria-owns="autosuggest__results"
         aria-activedescendant
         aria-haspopup="false"
         aria-expanded="false"
         id="autosuggest__input"
         initialvalue
         oninputchange="function onInputChange() {}"
         placeholder="Type 'G'"
         name="q"
         value
         class="form-control"
  >
  <div class="autosuggest__results-container">
  </div>
</div>

`;

exports[`Autosuggest can render default suggestion value by property name 1`] = `

<div id="autosuggest-25"
     data-server-rendered="true"
>
  <input type="text"
         autocomplete="off"
         role="combobox"
         aria-autocomplete="list"
         aria-owns="autosuggest__results"
         aria-activedescendant="autosuggest__results_item-0"
         aria-haspopup="true"
         aria-expanded="true"
         id="autosuggest__input"
         initialvalue
         oninputchange="function onInputChange() {}"
         placeholder="Type 'G'"
         name="q"
         onblur="function blurred() {
                mockFn();
              }"
         onfocus="function focused() {
                mockFn();
              }"
         value="Frodo"
         class="form-control autosuggest__input-open cool-class"
  >
  <div class="autosuggest__results-container">
    <div aria-labelledby="autosuggest-25"
         class="autosuggest__results"
    >
      <ul role="listbox"
          aria-labelledby="autosuggest"
      >
        <li role="option"
            data-suggestion-index="0"
            data-section-name="default"
            id="autosuggest__results_item-0"
            class="autosuggest__results_item"
        >
          undefined
        </li>
      </ul>
    </div>
  </div>
</div>

`;

exports[`Autosuggest can render simplest component with single onSelected 1`] = `

<div id="autosuggest-19"
     data-server-rendered="true"
>
  <input type="text"
         autocomplete="off"
         role="combobox"
         aria-autocomplete="list"
         aria-owns="autosuggest__results"
         aria-activedescendant
         aria-haspopup="true"
         aria-expanded="true"
         id="autosuggest__input"
         initialvalue
         oninputchange="function onInputChange() {}"
         placeholder="Type 'G'"
         name="q"
         value="phonics"
         class="form-control autosuggest__input-open cool-class"
  >
  <div class="autosuggest__results-container">
    <div aria-labelledby="autosuggest-19"
         class="autosuggest__results"
    >
      <ul role="listbox"
          aria-labelledby="autosuggest"
      >
        <li role="option"
            data-suggestion-index="0"
            data-section-name="default"
            id="autosuggest__results_item-0"
            class="autosuggest__results_item"
        >
          clifford kits
        </li>
        <li role="option"
            data-suggestion-index="1"
            data-section-name="default"
            id="autosuggest__results_item-1"
            class="autosuggest__results_item"
        >
          friendly chemistry
        </li>
        <li role="option"
            data-suggestion-index="2"
            data-section-name="default"
            id="autosuggest__results_item-2"
            class="autosuggest__results_item"
        >
          phonics
        </li>
        <li role="option"
            data-suggestion-index="3"
            data-section-name="default"
            id="autosuggest__results_item-3"
            class="autosuggest__results_item"
        >
          life of fred
        </li>
        <li role="option"
            data-suggestion-index="4"
            data-section-name="default"
            id="autosuggest__results_item-4"
            class="autosuggest__results_item"
        >
          life of fred math
        </li>
        <li role="option"
            data-suggestion-index="5"
            data-section-name="default"
            id="autosuggest__results_item-5"
            class="autosuggest__results_item"
        >
          magic school bus
        </li>
        <li role="option"
            data-suggestion-index="6"
            data-section-name="default"
            id="autosuggest__results_item-6"
            class="autosuggest__results_item"
        >
          math mammoth light blue
        </li>
        <li role="option"
            data-suggestion-index="7"
            data-section-name="default"
            id="autosuggest__results_item-7"
            class="autosuggest__results_item"
        >
          handwriting
        </li>
        <li role="option"
            data-suggestion-index="8"
            data-section-name="default"
            id="autosuggest__results_item-8"
            class="autosuggest__results_item"
        >
          math
        </li>
        <li role="option"
            data-suggestion-index="9"
            data-section-name="default"
            id="autosuggest__results_item-9"
            class="autosuggest__results_item"
        >
          minecraft
        </li>
        <li role="option"
            data-suggestion-index="10"
            data-section-name="default"
            id="autosuggest__results_item-10"
            class="autosuggest__results_item"
        >
          free worksheets
        </li>
        <li role="option"
            data-suggestion-index="11"
            data-section-name="default"
            id="autosuggest__results_item-11"
            class="autosuggest__results_item"
        >
          4th grade
        </li>
        <li role="option"
            data-suggestion-index="12"
            data-section-name="default"
            id="autosuggest__results_item-12"
            class="autosuggest__results_item"
        >
          snap circuits
        </li>
        <li role="option"
            data-suggestion-index="13"
            data-section-name="default"
            id="autosuggest__results_item-13"
            class="autosuggest__results_item"
        >
          bath toys
        </li>
        <li role="option"
            data-suggestion-index="14"
            data-section-name="default"
            id="autosuggest__results_item-14"
            class="autosuggest__results_item"
        >
          channies
        </li>
        <li role="option"
            data-suggestion-index="15"
            data-section-name="default"
            id="autosuggest__results_item-15"
            class="autosuggest__results_item"
        >
          fred
        </li>
        <li role="option"
            data-suggestion-index="16"
            data-section-name="default"
            id="autosuggest__results_item-16"
            class="autosuggest__results_item"
        >
          lego
        </li>
        <li role="option"
            data-suggestion-index="17"
            data-section-name="default"
            id="autosuggest__results_item-17"
            class="autosuggest__results_item"
        >
          math life of fred
        </li>
        <li role="option"
            data-suggestion-index="18"
            data-section-name="default"
            id="autosuggest__results_item-18"
            class="autosuggest__results_item"
        >
          multiplication
        </li>
        <li role="option"
            data-suggestion-index="19"
            data-section-name="default"
            id="autosuggest__results_item-19"
            class="autosuggest__results_item"
        >
          thinking tree
        </li>
      </ul>
    </div>
  </div>
</div>

`;

exports[`Autosuggest can render suggestions 1`] = `

<div id="autosuggest-2"
     data-server-rendered="true"
>
  <input type="text"
         autocomplete="off"
         role="combobox"
         aria-autocomplete="list"
         aria-owns="autosuggest__results"
         aria-activedescendant="autosuggest__results_item-0"
         aria-haspopup="true"
         aria-expanded="true"
         id="autosuggest__input"
         initialvalue
         oninputchange="function onInputChange() {}"
         placeholder="Type 'G'"
         onclick="function clicked() {
        mockFn();
      }"
         name="q"
         value="clifford kits"
         class="form-control autosuggest__input-open"
  >
  <div class="autosuggest__results-container">
    <div aria-labelledby="autosuggest-2"
         class="autosuggest__results"
    >
      <ul role="listbox"
          aria-labelledby="autosuggest"
      >
        <li role="option"
            data-suggestion-index="0"
            data-section-name="default"
            id="autosuggest__results_item-0"
            class="autosuggest__results_item"
        >
          clifford kits
        </li>
        <li role="option"
            data-suggestion-index="1"
            data-section-name="default"
            id="autosuggest__results_item-1"
            class="autosuggest__results_item"
        >
          friendly chemistry
        </li>
        <li role="option"
            data-suggestion-index="2"
            data-section-name="default"
            id="autosuggest__results_item-2"
            class="autosuggest__results_item"
        >
          phonics
        </li>
        <li role="option"
            data-suggestion-index="3"
            data-section-name="default"
            id="autosuggest__results_item-3"
            class="autosuggest__results_item"
        >
          life of fred
        </li>
        <li role="option"
            data-suggestion-index="4"
            data-section-name="default"
            id="autosuggest__results_item-4"
            class="autosuggest__results_item"
        >
          life of fred math
        </li>
      </ul>
    </div>
  </div>
</div>

`;

exports[`Autosuggest can select from suggestions using keystroke 1`] = `

<div id="autosuggest-7"
     data-server-rendered="true"
>
  <input type="text"
         autocomplete="off"
         role="combobox"
         aria-autocomplete="list"
         aria-owns="autosuggest__results"
         aria-activedescendant
         aria-haspopup="false"
         aria-expanded="false"
         id="autosuggest__input"
         initialvalue
         oninputchange="function onInputChange() {}"
         placeholder="Type 'G'"
         name="q"
         value="G"
         class="form-control"
  >
  <div class="autosuggest__results-container">
  </div>
</div>

`;

exports[`Autosuggest can use escape key to exit 1`] = `

<div id="autosuggest-5"
     data-server-rendered="true"
>
  <input type="text"
         autocomplete="off"
         role="combobox"
         aria-autocomplete="list"
         aria-owns="autosuggest__results"
         aria-activedescendant
         aria-haspopup="false"
         aria-expanded="false"
         id="autosuggest__input"
         initialvalue
         oninputchange="function onInputChange() {}"
         placeholder="Type 'G'"
         name="q"
         value="G"
         class="form-control"
  >
  <div class="autosuggest__results-container">
  </div>
</div>

`;

exports[`Autosuggest is aria complete 1`] = `

<div id="autosuggest-16"
     data-server-rendered="true"
>
  <input type="text"
         autocomplete="off"
         role="combobox"
         aria-autocomplete="list"
         aria-owns="autosuggest__results"
         aria-activedescendant="autosuggest__results_item-2"
         aria-haspopup="true"
         aria-expanded="true"
         id="autosuggest__input"
         initialvalue
         oninputchange="function onInputChange() {}"
         placeholder="Type 'G'"
         name="q"
         value="phonics"
         class="form-control autosuggest__input-open"
  >
  <div class="autosuggest__results-container">
    <div aria-labelledby="autosuggest-16"
         class="autosuggest__results"
    >
      <ul role="listbox"
          aria-labelledby="autosuggest"
      >
        <li role="option"
            data-suggestion-index="0"
            data-section-name="default"
            id="autosuggest__results_item-0"
            class="autosuggest__results_item"
        >
          clifford kits
        </li>
        <li role="option"
            data-suggestion-index="1"
            data-section-name="default"
            id="autosuggest__results_item-1"
            class="autosuggest__results_item"
        >
          friendly chemistry
        </li>
        <li role="option"
            data-suggestion-index="2"
            data-section-name="default"
            id="autosuggest__results_item-2"
            class="autosuggest__results_item"
        >
          phonics
        </li>
        <li role="option"
            data-suggestion-index="3"
            data-section-name="default"
            id="autosuggest__results_item-3"
            class="autosuggest__results_item"
        >
          life of fred
        </li>
        <li role="option"
            data-suggestion-index="4"
            data-section-name="default"
            id="autosuggest__results_item-4"
            class="autosuggest__results_item"
        >
          life of fred math
        </li>
      </ul>
    </div>
  </div>
</div>

`;

exports[`Autosuggest onBlur and onFocus work as expected, including deprecation warnings 1`] = `

<div id="autosuggest-22"
     data-server-rendered="true"
>
  <input type="text"
         autocomplete="off"
         role="combobox"
         aria-autocomplete="list"
         aria-owns="autosuggest__results"
         aria-activedescendant
         aria-haspopup="true"
         aria-expanded="true"
         id="autosuggest__input"
         initialvalue
         oninputchange="function onInputChange() {}"
         placeholder="Type 'G'"
         name="q"
         onblur="function blurred() {
                mockFn();
              }"
         onfocus="function focused() {
                mockFn();
              }"
         value="G"
         class="form-control autosuggest__input-open"
  >
  <div class="autosuggest__results-container">
    <div aria-labelledby="autosuggest-22"
         class="autosuggest__results"
    >
      <ul role="listbox"
          aria-labelledby="autosuggest"
      >
        <li role="option"
            data-suggestion-index="0"
            data-section-name="default"
            id="autosuggest__results_item-0"
            class="autosuggest__results_item"
        >
          clifford kits
        </li>
        <li role="option"
            data-suggestion-index="1"
            data-section-name="default"
            id="autosuggest__results_item-1"
            class="autosuggest__results_item"
        >
          friendly chemistry
        </li>
        <li role="option"
            data-suggestion-index="2"
            data-section-name="default"
            id="autosuggest__results_item-2"
            class="autosuggest__results_item"
        >
          phonics
        </li>
        <li role="option"
            data-suggestion-index="3"
            data-section-name="default"
            id="autosuggest__results_item-3"
            class="autosuggest__results_item"
        >
          life of fred
        </li>
        <li role="option"
            data-suggestion-index="4"
            data-section-name="default"
            id="autosuggest__results_item-4"
            class="autosuggest__results_item"
        >
          life of fred math
        </li>
      </ul>
    </div>
  </div>
</div>

`;
